cmd/fts/cmd/tasks.go

```diff
@@ -0,0 +1,232 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	client "github.com/textileio/powergate/api/client"
+	userPb "github.com/textileio/powergate/api/gen/powergate/user/v1"
+)
+
+type Task struct {
+	name     string
+	path     string
+	bytes    int64
+	isDir    bool
+	isHidden bool
+}
+
+type TaskResult struct {
+	task    Task
+	cid     string
+	jobId   string
+	stage   string
+	state   *client.WatchStorageJobsEvent
+	records []*userPb.StorageDealRecord
+	err     error
+}
+
+type RunConfig struct {
+	token           string
+	ipfsrevproxy    string
+	serverAddress   string
+	concurrent      int64
+	minDealBytes    int64
+	maxStagingBytes int64
+	dryRun          bool
+}
+
+func Run(tasks []Task, config RunConfig) chan TaskResult {
+
+	sc := StagingConfig{
+		maxStagingBytes: config.maxStagingBytes,
+		minDealBytes:    config.minDealBytes,
+	}
+
+	// wg is used to force the application wait for all goroutines to finish before exiting.
+	wg := sync.WaitGroup{}
+	// ch is a buffered channel for processing tasks
+	ch := make(chan bool, config.concurrent)
+
+	// res is a buffered channel for receiving results
+	res := make(chan TaskResult, len(tasks))
+
+	// How many jobs to expect
+	wg.Add(len(tasks))
+
+	for _, task := range tasks {
+		go ProcessTask(task, config, ch, res, &wg, &sc)
+	}
+
+	go func() {
+		wg.Wait()
+		close(res)
+	}()
+
+	return res
+}
+
+func ProcessTask(task Task, config RunConfig, ch chan bool, res chan<- TaskResult, wg *sync.WaitGroup, sc *StagingConfig) {
+	defer wg.Done()
+
+	// Fill in the channel buffer slot.
+	ch <- true
```

Comment:
I think you'd like to do `defer func() { <- ch}` in the line below. That will get you covered of returning the slot in any code path that happens in the method. Maybe you were careful enough to `break` and be sure of never early returning, but if you do the above, you can forget about messing around with that situation. Usually safer when some other person comes to change the code and doesn't have that in mind.